### ISSUE TITLE
Fold `purescript-psa` into `purs compile`

### ISSUE THREAD

**Issue by JordanMartinez** [purescript/purescript#4143]
## Summary

Update the compiler to do the same tasks done by [`purescript-psa`](https://github.com/natefaubion/purescript-psa), so that one does not need to have `purescript-psa` installed.

## Motivation

This idea has already been mentioned and found support from both [Harry](https://github.com/purescript/purescript/issues/3778#issuecomment-578986966) and [Nate](https://github.com/purescript/purescript/issues/3778#issuecomment-578987087), the author of `purescript-psa`. However, it's only now being formally proposed.

AFAIK, there are two things the compiler cannot do that requires `purescript-psa` to implement:
- returning a non-zero exit code (e.g. failing the CI build) if there are any warnings (via the `--strict` flag)
- censoring and/or filtering errors and/or warnings (via the `censor-codes` and `filter-codes` args)

In addition, `purescript-psa` outputs information in a more helpful manner than `purs compile`'s default:
- colored warnings and errors, which makes it easier to see which are warnings and which are errors.
- statistics on the number of warnings and errors across both the library directory and source directory

When we were migrating the ecosystem to `v0.14.0`, we had to use `purescript-psa` at some point to do what the compiler was unable to do: fail the CI build if a warning was produced. However, this complicated things as well when [`MonadZero` was deprecated](https://github.com/purescript/purescript-control/issues/62), as we wanted to continue emitting warnings about the type class' deprecation. 

Other related issues where `purescript-psa` provided solutions not yet implemented in the compiler: #1898 and #2308

## Proposal

I propose we port `purescript-psa` to the compiler, so that what one would normally run via `purescript-psa` would work via `purs compile`. Below is the relevant args that `purescript-psa` provides per its ['usage' docs](https://github.com/natefaubion/purescript-psa/blob/master/src/Main.purs#L314-L343):
```
  --verbose-stats        Show counts for each warning type
  --censor-stats         Censor warning/error summary
  --censor-warnings      Censor all warnings
  --censor-lib           Censor warnings from library sources
  --censor-src           Censor warnings from project sources
  --censor-codes=CODES   Censor specific error codes
  --filter-codes=CODES   Only show specific error codes
  --no-colors            Disable ANSI colors
  --no-source            Disable original source code printing
  --strict               Promotes src warnings to errors
  --stash                Enable persistent warnings (defaults to .psa-stash)
  --stash=FILE           Enable persistent warnings using a specific stash file
  --is-lib=DIR           Distinguishing library path (defaults to 'bower_components' and `.spago`)

  CODES                  Comma-separated list of purs error codes
```

The easy thing to do would be to port these args directly to `purs compile` with no modifications. This would maintain compatibility with `purescript-psa`, so that tools like `spago` and `pulp` only have to change how they call `purs`.

## Other Questions

While maintaining compatibility is desirable, is this the design we would want to use for this feature? There are two other features that could be implemented in this process that break away from the approach followed by `purescript-psa`. Depending on how these are implemented, we might want to change the CLI to better support them. Rather than implementing this feature so that it's compatible with `purescript-psa` and then creating a breaking change later to support the below features, it might be better to implement this idea with the below features already in mind.

Feature 1: Library-specific and/or module-specific warning/error code filtering and/or censoring

Right now, one can censor/filter codes in all libraries found in the library directory (e.g. `.spago` or `bower_components`), in the source directory (e.g. `src`), or in both. However, one cannot censor/filter codes on an individual library and/or module level. If we decided to support this feature, we may want to change how that information is passed into `purs compile`. 

Note: this idea is essentially natefaubion/purescript-psa#2 and natefaubion/purescript-psa#20

Feature 2: Finer-grained code censoring/filtering based on what the error message is

Right now, any censoring/filtering of a code is an all-or-nothing approach. One cannot censor specific codes if they match some criteria (e.g. [a "UserDefined" warning code with the "`MonadZero` is deprecated" message](https://github.com/purescript/purescript-control/issues/62#issuecomment-737766403)). 

## How would this affect tooling?

For context, the current release is `v0.14.2` with a `v0.14.3` coming soon. Let's say for argument's sake that this feature was released in a hypothetical `v0.19.0` release. `spago` and `pulp` are front-ends to `purs` used to build code. In all releases before `v0.19.0`, they would need to use `purescript-psa` to as a front-end to calling `purs` in case one wanted to use one of its flags (e.g. `--strict`). In `v0.19.0` and every release thereafter, they would call `purs` directly with such arguments. 

If we implement the above two features, it would complicate the tooling a bit as they would need to support both approaches for a time.

**Comment by rhendric** [comment]
:+1: to folding some aspects in (especially `--strict`), but :-1: to simply porting the functionality directly without more analysis.

What are the list of things PureScript warns about? What are the actual user interaction stories with these warnings that would make it useful to suppress them? Would those stories be better served by adding additional data to warnings, instead of filtering based on the data that are already there?

Here's an example story: Alice has a PureScript project that compiles with no warnings on PS 0.19.5. A dependency pushed her to upgrade to 0.20.0, which caused several new warnings to be added, along with a few breaking changes. Alice wants to get her project working as quickly as possible, but doesn't want to relax her standards for warnings that were already present in 0.19.5. So Alice might be best served by filtering warnings *based on the compiler version in which they were introduced*.

Another: Bailey isn't upgrading their compiler, but they did just upgrade one of their dependencies which deprecated some parts of its API. Bailey isn't ready to rewrite the code depending on that API yet, so they would like to filter warnings *based on the source path in which the warning was declared* (as opposed to the path in which the warning is emitted).

Another: Colin just doesn't think PureScript should be doing linting. He wants any current and future deprecation warnings to fail his build, but any current and future linting warnings should be ignored, and he doesn't want to maintain a list of what those codes are. Colin wants to filter warnings *based on their category*.

I'm not making an argument for infinite configurability, to be clear. I think we should find the highest-impact knobs and implement those, and if those just *happen* to be the knobs that purescript-psa currently uses, then great! But I wouldn't expect that to be the case.

**Comment by JordanMartinez** [comment]
At the very least, I think implementing `--strict` should be fairly straightforward. So, that could get in as its own simple PR. It would take more thought for the other things (which I agree is a better approach to take initially).

Restating what you've mentioned above, `purs compile` builds code, but its the contexts in which it is called that matter:
- before/after updating a dependency
- before/after updating the compiler

The Colin example above is an example of user-preference configuration.

> So Alice might be best served by filtering warnings based on the compiler version in which they were introduced.

Are there ever times when we might need to update an error added in previous versions because of a new feature added? If so, should this instead say, "based on the compiler version in which they were introduced or last modified?"

**Comment by rhendric** [comment]
> Are there ever times when we might need to update an error added in previous versions because of a new feature added? If so, should this instead say, "based on the compiler version in which they were introduced or last modified?"

Probably, yeah. This might be difficult in practice to get exactly right, because the change that causes the warning to be added might be in a completely different part of the compiler from where the warning is generated. But in the ideal design space, yes, I think that's right for that use case.

**Comment by hdgarrood** [comment]
I think there are number of tasks here which should probably remain as separate issues, especially if there are some design questions to work through. Perhaps we should start with the parts which we are least likely to disagree on the design of:

- #1898
- #2308
- Add a `--strict` flag which makes src warnings into errors
- Show stats at the end of compilations

and then start thinking about controlling which warnings are shown later.

**Comment by JordanMartinez** [comment]
I'll open an issue for the above 3 that don't have one.

**Comment by JordanMartinez** [comment]
I feel like we've come full-circle. Originally, I opened this issue so that errors/warnings could be improved, so that one did not have to use `psa`. Harry encouraged me to tackle each issue individually, but in the process of doing so, it seemed that Ryan pointed out that these issues are interrelated. In other words, they must be designed together.

Due to the Spago rewrite in PureScript, I started porting `psa`/`psa-utils` into the new Spago when it calls `purs compile` here: https://github.com/purescript/spago/pull/955. However, `psa`/`psa-utils` are used by both `spago` and the IDE. If one gets updated without the other being updated, we'll just have a different problem.

All that to say, I think this request should be seen as a single feature: improving the compiler's outputted warnings/errors when `purs compile` is used, regardless of whether it be called directly, via `spago`, or the IDE.

Thoughts?

**Comment by JordanMartinez** [comment]
<details>

<summary>Or put differently, here is the control flow of `psa` and how these features are interrelated:</summary>

```mermaid
flowchart TD
  subgraph errorWarnings0 ["Compiler warning/error info"]
    warnings0["Warnings (initial)"]
    errors0["Errors (initial)"]
  end
  sourceOrLib{"Is glob a\nsource or lib\npackage?"}
  subgraph censorConfig ["Censor configuration"]
    censorByGlob("Censor by \nglob (e.g. \n--censor-lib\n--censor-src\n)")
    censorByCode("Censor by \ncode (e.g. \n--filter-code)")
    subgraph censorOther ["Censor by some future thing"]
      censorByCategory("Censor by error category")
    end
  end
  
  warnings0 -->|"are considered by the censor config..."| censorConfig
  censorConfig -->|"... and produce"| censoredWarnings1
  censorConfig -->|"... and produce"| uncensoredWarnings1
  sourceOrLib --> |"Indicates which globs are which"| censorByGlob
  errors0 --> errors1

  subgraph errorWarnings1 ["Compiler warning/error info"]
    censoredWarnings1["Warnings (censored)"]
    uncensoredWarnings1["Warnings (uncensored)"]
    errors1["Errors (initial)"]
  end

  strict{"Consider warnings\nas errors?\n(--strict)"}

  censoredWarnings1 --> censoredWarnings2a
  censoredWarnings1 --> censoredWarnings2b
  uncensoredWarnings1 --> strict
  strict -->|"Leave as is"| uncensoredWarnings2a
  strict -->|"Promote warnings to errors"| errorWarnings2b
  errors1 --> errors2a
  errors1 --> errors2b

  subgraph errorWarnings2a ["Compiler warning/error info"]
    censoredWarnings2a["Warnings (censored)"]
    uncensoredWarnings2a["Warnings (uncensored)"]
    errors2a["Errors (initial)"]
  end

  subgraph errorWarnings2b ["Compiler warning/error info"]
    censoredWarnings2b["Warnings (censored)"]
    errors2b["Errors (initial + promoted uncensored warnings)"]
  end
  
  subgraph stats ["Provide statistical summary"]
    none["No stats"]

    subgraph compactStats ["Compact Stats"]
      summary1["      Src     Lib    Total\nWarning   4   2    8\nError    1    0     1\nTotal    5    2     9"]
    end

    subgraph verboseStats ["Verbose Stats"]
      summary2["      Src     Lib    Total\nCode1   4   2    8\nCode2    1    0     1\nCode3    1    0     1\nCode4    1    0     1\nTotal    5    2     11"]
    end
  end

  errorWarnings2a --> stats
  errorWarnings2b --> stats
```

</details>

> All that to say, I think this request should be seen as a single feature: improving the compiler's outputted warnings/errors when purs compile is used, regardless of whether it be called directly, via spago, or the IDE.

Having looked through more of these issues, I think things can still be tackled as separate issues, but I would currently group things together in this way:
- `psa`'s `--ansi` option - #4152 / #3778 can still be its own issue because it's just adding colors to current output when not using `--json-errors`.
- `psa`'s `--stash` option - #2105, #2308, and #3198 are about persisting compiler warnings across builds. While one can use a tool like `psa`'s `--stash` option, the issue is that this information gets out-of-sync.
- `psa` include source code snippet in message - #1898 can still be its own issue. 
- `psa` allows censoring warnings by glob/code and outputs stats - #4153, #4154, and #1923 should still be considered together.
    - #4154 is dependent on whether warnings can be censored (#1923) and whether warnings are promotable to errors (#4153).
    - Most notably, #1923 provides insight as to why `psa`'s `censor-code` likely exists: because some of our warnings are more like lint warnings rather than actual warnings. So, implementing this feature may require distinguishing our current warnings into three categories: error, warnings, and lintings. But, if the new Spago rewrite allows one to censor warnings by code in the new configuration file, then perhaps this distinction isn't an issue? And while an external `pslint` tool would be desirable, it doesn't currently exist. If it was written in PureScript, I think it's blocked by https://github.com/natefaubion/purescript-language-cst-parser/issues/28

**Comment by rhendric** [comment]
So, I don't think all of these features are necessarily interrelated. I think censoring warnings is related to making the compiler package-aware and both of those features should be designed together, ideally starting from the perspective of users in various scenarios who want `purs` to do the best thing for them in those circumstances (as opposed to starting with a set of capabilities that `psa` currently provides, or that would be easy for us to provide, or that it's easy to imagine how they would work). I think it'd be basically harmless to implement your first three bullet points without diving into that bramble patch (`--ansi`, `--stash`, and including source in errors). I also think `--strict` can be detached from the category of ‘censoring warnings’ and tackled on its own.

But let me ask these questions:
* With `purs` as it is today, what toolchain would you recommend to a new PureScript user for the best out-of-the-box experience, and does it include `psa`?
* Does that answer change if we implement `--ansi` in the compiler?
* Does that answer change if we implement `--stash` in the compiler?
* Does that answer change if we implement code excerpts in the compiler?
* Does that answer change if we implement `--strict` in the compiler?

My answers are that I would recommend `psa` and probably keep recommending `psa` until the warning censoring thing is figured out, regardless of those other features because `psa` currently provides those other features. Maybe your answers are different, and I'd love to hear why if so.

But if your answers are the same as mine, then why would you—why would any of us—want to spend time working on those other features in the compiler before spending time working out even a plan for figuring out the warning censoring thing? Because that's the *highest-risk* part of all of this, and without it what value does the rest of the work deliver? To me, it seems like the [politician's syllogism](https://en.wikipedia.org/wiki/Politician%27s_syllogism) at work to ignore that high-risk component and just work on the easier chunks.

I could be missing something in the above analysis and I am not vetoing any work on any of the above except for working on directly porting warning censorship from `psa` without thinking about how it interacts with related proposed compiler features. If this argument doesn't feel right to you go right ahead and implement all those other features! I'm seriously not trying to be the says-no-to-progress-guy here. I'm just trying to direct our efforts toward the most impactful progress, and I might be wrong.

**Comment by natefaubion** [comment]
`—strict` is only useful if you can censor library warnings. Its primary use case is enforcing warning free code in CI. Without lib censoring, libraries may fail your CI due to warnings.

**Comment by JordanMartinez** [comment]
> I'm seriously not trying to be the says-no-to-progress-guy here. I'm just trying to direct our efforts toward the most impactful progress, and I might be wrong.

Understood, and I do appreciate your feedback here.

> With purs as it is today, what toolchain would you recommend to a new PureScript user for the best out-of-the-box experience, and does it include psa?

Yes, it would include `psa`.
- I recommend it here: https://jordanmartinez.github.io/purescript-jordans-reference-site/content/01-Getting-Started/04-Install-Guide.html#overview
- It's been recommended for the last four years in these Discourse posts done by Thomas about the recommending tooling for PureScript:
    - [2022](https://discourse.purescript.org/t/recommended-tooling-for-purescript-in-2022/3206)
    - [2021](https://discourse.purescript.org/t/recommended-tooling-for-purescript-in-2021/2725)
    - [2020](https://discourse.purescript.org/t/recommended-tooling-for-purescript-in-2020/1615)
    - [2019](https://discourse.purescript.org/t/recommended-tooling-for-purescript-applications-in-2019/948)

> Does that answer change if we implement --ansi in the compiler?

No. Colors are nice, but it's not "enough"

> Does that answer change if we implement --stash in the compiler?

No. I appreciate knowing what the warnings are so I can fix them rather than having to write `rm -rf output/Module.Name && spago build` just to see what the warnings are / whether there are any warnings. However, it's still not enough.

> Does that answer change if we implement code excerpts in the compiler?

No. While this is helpful, the IDE allows me to jump to those locations anyways. When the IDE crashes or is unreliable, then this would be helpful.

> Does that answer change if we implement --strict in the compiler?

Somewhat yes but only if it implemented the warning censoring. This feature allows me to use CI to ensure my project's code is warning-free while allowing me to use libraries that may still emit warnings (see Nate's comment above). However, I would also want the other features listed above. For example, the ansi helps me distinguish between problems I must fix (errors) from problems I can fix later (warnings). And when I'm skimming through the output of a `purs compile` call, I can immediately see when I can stop scrolling up based on colors alone.

All this to say, I don't think `--strict` can be detached from warning censoring.

> I think censoring warnings is related to making the compiler package-aware and both of those features should be designed together

I'll need to re-read through that issue, but yes you have a point here. And in some ways, the Spago rewrite intends to deal with some of these problems itself.

**Comment by JordanMartinez** [comment]
In the event that supporting `--strict` can only be done by making the compiler package-aware, and if that latter task is a thorny one, then I think integrating `psa` into `spago` is a good temporary solution so long as the underlying issue that gives rise to `psa`'s `--stash` flag is resolved.
